### PR TITLE
added bitcoin

### PR DESCRIPTION
**Pull request by elRaulito**
**NOTE**

#### Please enable "Allow edits by maintainers" while putting up the PR.

---

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
4. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
5. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
6. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---

### CODE REVIEW

**Comment by llamatester** [comment]
Error while running adapter at projects/fluidtokens: 
        
 
  
    
    
    SyntaxError: Unexpected token ';'
    Truncated error stack:
    
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:127:12
        at Object.<anonymous> (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:259:3)
    Error saving cache: Error: ENOENT: no such file or directory, open '/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@defillama/sdk/build/util/local_cache/zlib-1.0/BlockCache.json'

**Comment by llamatester** [comment]
Error while running adapter at projects/fluidtokens: 
        
 
  
    
    
    Error: 
        Found export keys that were not part of specification: bitcoin_tvl
    
        List of valid keys: 
    				
    				tvl
    				staking
    				methodology
    				pool2
    				misrepresentedTokens
    				fetch
    				timetravel
    				borrowed
    				start
    				doublecounted
    				hallmarks
    				isHeavyProtocol
    				deadFrom
    				ownTokens
    				vesting
        
    Error saving cache: Error: ENOENT: no such file or directory, open '/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@defillama/sdk/build/util/local_cache/zlib-1.0/BlockCache.json'

**Comment by llamatester** [comment]
The adapter at projects/fluidtokens exports TVL: 
        
 
 
    cardano-staking           4.89 M
    staking                   4.89 M
    bitcoin                   1.57 M
    cardano                   1.57 M
    cardano-borrowed          691.23 k
    borrowed                  691.23 k
    
    total                    3.15 M

**Comment by elRaulito** [comment]
@g1nt0ki I think I am doing something wrong here, I should have around 500k $ of Bitcoin TVL from the API, but I think I am not defining it in the right way now that I added bitcoin

**Review by g1nt0ki — CHANGES_REQUESTED** [review]
no, this will duplicate tvl between bitcoin & cardano

you need to create a new tvl function for bitcoin, also, for bitcoin, we need the staking addresses, we dont accept api for it, you can add your protocol addresses here: 

https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/helper/bitcoin-book/index.js

and use it like: https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/chakra/index.js#L6

**Comment by llamatester** [comment]
The adapter at projects/fluidtokens exports TVL: 
        
 
 
    cardano-staking           4.92 M
    staking                   4.92 M
    cardano                   1.08 M
    cardano-borrowed          707.18 k
    borrowed                  707.18 k
    bitcoin                   514.35 k
    
    total                    1.59 M

**Comment by elRaulito** [comment]
@g1nt0ki managed to fix, thanks a lot

However we have a not-custodial time lock script, so every user has his own address, like shown at this link

https://api2.fluidtokens.com/get-stake-amount-addresses?nameStakeVault=RAREPUPSPUPCASH

We are similar to what Babylon is doing 

https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/babylon/index.js

and I see they also use a API, probably for the same reason, many addresses to track

**Comment by g1nt0ki** [comment]
hi @elRaulito we missed babylon, and currently pushing them to share bitcoin address list like the rest of the protocols. 

Thought a bit more about this, think it is better to create new listing for bitcoin staking part. then we will create parent listing for you, like we have for aave: https://defillama.com/protocol/aave

---
This is just for my curiosity, how does the locking work? it is not like bitcoin supports proper scripting/smart contracts

**Comment by elRaulito** [comment]
@g1nt0ki completly missed the question

- Bitcoin does not have smart contracts however has time lock scripts
- Users can lock a certain asset until a certain date under their key
- This makes it not custodial because they will be the only ones able to spend and also creates a lock time staking structure
- Where do rewards come from? Projects, revenue share or just airdrops

Tldr: Bitcoin allow to say I won't spend this until X date, project can verify it onchain and reward users that do it

**Comment by llamatester** [comment]
The adapter at projects/fluidtokens exports TVL: 
        
 
 
    cardano-staking           4.90 M
    staking                   4.90 M
    cardano                   1.09 M
    cardano-borrowed          666.34 k
    borrowed                  666.34 k
    bitcoin                   400.57 k
    
    total                    1.49 M

**Comment by g1nt0ki** [comment]
> @g1nt0ki completly missed the question
> 
>     * Bitcoin does not have smart contracts however has time lock scripts
> 
>     * Users can lock a certain asset until a certain date under their key
> 
>     * This makes it not custodial because they will be the only ones able to spend and also creates a lock time staking structure
> 
>     * Where do rewards come from? Projects, revenue share or just airdrops
> 
> 
> Tldr: Bitcoin allow to say I won't spend this until X date, project can verify it onchain and reward users that do it

got it. still feels like it is a different product to what we have listed now, can you create a new adapter and compute tvl on chain using calls to bitcoin chain, you can use this method to compute brc20 token tvl: https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/helper/chain/brc20.js#L6

**Comment by elRaulito** [comment]
@g1nt0ki won't it eventually go in time-out?  we have more or less 1000 users, each of them has their address to stake and we should compute the sum of the TVL

We can even provide the list of all these addresses and the tvl of each of them if that's required, we just don't want to slow down on your side

**Comment by g1nt0ki** [comment]
another reason why we prefer the address and balance list is, we want to price tokens using our coins api, this way all projects use the same token price and users can see token breakdown of a project

**Comment by waynebruce0x** [comment]
closing for inactivity, please let us know if youd like to continue with this further
